File: strings/541-reverse-string-ii.java
class Solution {
    public String reverseStr(String s, int k) {
        StringBuilder result = new StringBuilder();
        int length = s.length();
        
        for (int i = 0; i < length; i += 2 * k) {
            // Reverse the first k characters in the current segment
            int end = Math.min(i + k, length); // Ensure we don't go past the end of the string
            String toReverse = s.substring(i, end);
            result.append(new StringBuilder(toReverse).reverse());
            
            // Append the next k characters as they are
            int startNextSegment = end;
            end = Math.min(i + 2 * k, length);
            if (startNextSegment < end) {
                result.append(s.substring(startNextSegment, end));
            }
        }
        
        return result.toString();
    }
}

File: strings/541-reverse-string-ii.py
class Solution:
    def reverseStr(self, s: str, k: int) -> str:

        parsed_s = []
        left = 0
        n = len(s)
        while left < n:
            right = min(left + 2*k, n)
            parsed_s.append(s[left:right])
            left = right

        rev_parsed_strings = []

        for string in parsed_s:
            rev = string[:k][::-1]
            new_str = rev + string[k:]
            rev_parsed_strings.append(new_str)

        result = ""
        
        for string in rev_parsed_strings:
            result += string

        return result
